plugins/ccgx/fu-ccgx-dmc-device.c

```diff
@@ -0,0 +1,703 @@
+			/* get reponse  */
+			for (guint8 retry = 0; retry < 2; retry++) {
+				g_autoptr(GError) error_local = NULL;
+				/* get interrupt request */
+				if (!fu_ccgx_dmc_device_read_intr_req (self, &dmc_int_req, &error_local)) {
+					g_set_error (error,
+						     FWUPD_ERROR,
+						     FWUPD_ERROR_INTERNAL,
+						     "failed read intr req: %s",
+						     error_local->message);
+				       return FALSE;
+				}
+				/* check opcode for fw write */
+				if (dmc_int_req.opcode != DMC_INT_OPCODE_IMG_WRITE_STATUS) {
+					g_set_error (error,
+						     FWUPD_ERROR,
+						     FWUPD_ERROR_NOT_SUPPORTED,
+						     "invalid dmc intr req opcode in fw write = %d",
+						     dmc_int_req.opcode);
+						return FALSE;
+				}
+				if (dmc_int_req.data[0] == 0) /* success */
+					break;
+				else if (dmc_int_req.data[0] != 1)  {/* retry if data[0] is 1 otherwise error */
+					g_set_error (error,
+						     FWUPD_ERROR,
+						     FWUPD_ERROR_NOT_SUPPORTED,
+						     "invalid dmc intr req data in fw write = %d",
+						     dmc_int_req.data[0]);
+						return FALSE;
+				}
+				g_usleep (DMC_FW_WRITE_STATUS_RETRY_DELAY_MS*1000);
+			}
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_dmc_write_firmware (FuDevice *device,
+			    FuFirmware *firmware,
+			    FwupdInstallFlags flags,
+			    GError **error)
+{
+	FuCcgxDmcDevice *self = FU_CCGX_DMC_DEVICE (device);
+	GBytes *fwct_record = NULL;
+	GBytes *custom_meta_record = NULL;
+	GPtrArray *image_records = NULL;
+	const guint8* custom_meta_data = NULL;
+	gsize custom_meta_size = 0;
+	const guint8* fwct_data = NULL;
+	gsize fwct_size = 0;
+	DmcIntRqt dmc_int_rqt = {0};
+	guint8 img_index = 0;
+	FuCcgxDmcFirmwareImageRecord *img_rcd = NULL;
+
+	/* get fwct record */
+	fwct_record = fu_ccgx_dmc_firmware_get_fwct_record (FU_CCGX_DMC_FIRMWARE(firmware));
+	/* get custom meta record */
+	custom_meta_record = fu_ccgx_dmc_firmware_get_custom_meta_record (FU_CCGX_DMC_FIRMWARE(firmware));
+	/* get image records */
+	image_records = fu_ccgx_dmc_firmware_get_image_records (FU_CCGX_DMC_FIRMWARE(firmware));
+
+	if (custom_meta_record)
+		custom_meta_data = g_bytes_get_data (custom_meta_record, &custom_meta_size);
+
+	g_debug ("reset dmc state machine");
+	if (!fu_ccgx_dmc_device_send_reset_state_machine (self, error))
+		return FALSE;
+
+	/* start fw upgrade with custom meta data */
+	if (!fu_ccgx_dmc_device_send_start_upgrade (self, custom_meta_data, custom_meta_size, error))
+		return FALSE;
+
+	fwct_data = g_bytes_get_data (fwct_record, &fwct_size);
+	if (fwct_data == NULL) {
+		g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid fwct data");
+		return FALSE;
+	}
+
+	/* send fwct data */
+	g_debug ("fw write request with meta:%u fwct:%u", (guint32)custom_meta_size, (guint32)fwct_size);
+	if (!fu_ccgx_dmc_device_send_fwct (self,fwct_data, fwct_size, error))
+		return FALSE;
+
+
+	/* init fw written */
+	self->fw_data_written = 0;
```

Comment:
no, `self` represents attributes about the device, not progress.

Reply:
Replace it with local variable on today commit (ccgx: fix code alignment, modify code for review)